subsys/net/lib/coap/coap_client.c

```diff
@@ -380,8 +389,11 @@ int coap_client_req(struct coap_client *client, int sock, const struct sockaddr
 static void report_callback_error(struct coap_client_internal_request *internal_req, int error_code)
 {
 	if (internal_req->coap_request.cb) {
-		internal_req->coap_request.cb(error_code, 0, NULL, 0, true,
-					      internal_req->coap_request.user_data);
+		if (!atomic_set(&internal_req->in_callback, 1)) {
+			internal_req->coap_request.cb(error_code, 0, NULL, 0, true,
+						      internal_req->coap_request.user_data);
+			atomic_clear(&internal_req->in_callback);
+		}
```

Comment:
Would an else block printing a debug message stating there was an error during handling of callback make sense? I worry a little about potentially causing dropped error callbacks. But I don't understand this client well enough to know if that concern is reasonable

Reply:
It will only skip reporting when the report is due to a call to the cancel function from the same callback that would receive the error notification. But I went ahead and made that explicit.